### ISSUE TITLE
Deterministic hang in to_torch depending on deallocate location

### ISSUE THREAD

**Issue by yieldthought** [tenstorrent/tt-metal#7822]
We've tracked down a deterministic hang with a very strange trigger based on _when_ an intermediate tensor is deallocated.

To reproduce:

Checkout 4c917440c47577050a4374f01caa2465616d2c8a and run:
`pytest models/demos/mistral7b/tests/test_mistral_perf.py::test_mistral_model_perf[32-15-0.16]`

## Hanging configuration

Here's where things get interesting. This is the configuration (as seen in commit xxx) that causes the hang:

Decoder:
```
attn_norm = self.attention_norm(x) # <- attn_norm is the problematic tensor
# Attention module expects a list of inputs, attn masks (multi-device support)
r = self.attention.forward(
    [attn_norm],
    current_pos,
    [attn_masks],
)
# ttnn.deallocate(attn_norm) <- attn_norm is deallocated inside attention.forward() in this version
```

Attention:
```
xqkv_fused = ttnn.linear(
    attn_norm,
    wqkv,
    program_config=self.wqkv_program_config,
    memory_config=self.model_config["XQKV_MM_OUTPUT_MEMCFG"],
    compute_kernel_config=self.compute_kernel_config,
    dtype=self.dtype,
)
 
ttnn.deallocate(attn_norm, force=False) # <- when deallocated here, the final to_torch on the model output hangs!
```

## Working configuration

Now, if instead we comment out the deallocate call in the attention block and instead deallocate attn_norm in the decoder after the attention block returns, _there is no hang_, e.g. in this configuration:

Decoder:
```
attn_norm = self.attention_norm(x) # <- attn_norm is the problematic tensor
# Attention module expects a list of inputs, attn masks (multi-device support)
r = self.attention.forward(
    [attn_norm],
    current_pos,
    [attn_masks],
)
ttnn.deallocate(attn_norm) # Deallocating here avoids the hang?!
```

Attention:
```
xqkv_fused = ttnn.linear(
    attn_norm,
    wqkv,
    program_config=self.wqkv_program_config,
    memory_config=self.model_config["XQKV_MM_OUTPUT_MEMCFG"],
    compute_kernel_config=self.compute_kernel_config,
    dtype=self.dtype,
)
 
# ttnn.deallocate(attn_norm, force=False) <- no longer deallocate here even though unused in the rest of this function
```

This is weird AF and I hope it helps someone debug the actual cause of this hang, which might also be presenting elsewhere in the stack but less deterministically.

**Comment by jliangTT** [comment]
high bw discussion here - https://tenstorrent.slack.com/archives/C065NUB85K3/p1714041186227089

**Comment by uaydonat** [comment]
@yieldthought @mtairum we have been asked to try this with fd2.0 in case with some weird interaction, it goes away

**Comment by yieldthought** [comment]
### FD 2.0 update

This still hangs on today's main, but now the workaround (moving the deallocate) also hangs, e.g. this code now hangs every time regardless of where the deallocate call is made from. The version pushed in `yieldthought/7822-fd2` is the configuration that would run cleanly on fd 1, but both this and the hang configuration hang on FD 2.

**Comment by uaydonat** [comment]
@yieldthought does it mean mistral completely is broken? If so, we need to raise this to P0

**Comment by jvasilje** [comment]
@arakhmati what's the status on this issue? this looks like a deallocate host issue?

**Comment by arakhmati** [comment]
@yieldthought I can't run this on my WH machine. I don't have `/mnt/MLPerf` or `/proj_sw`. I am not sure why both are missing. Maybe one of them can be mounted. But is it possible to make this test run without relying on any directory? And  just using `/tmp` directory or something like that to store the store the weights?

**Comment by arakhmati** [comment]
@jvasilje this issue looks like it's deeper in the stack. Can someone who has a better understanding of watcher be assigned to re-run this test and see what's causing the hang?

**Comment by yieldthought** [comment]
@uaydonat This only occurs on the new attention implementation branch and does not occur on the current main, so not a P0 imo.

**Comment by yieldthought** [comment]
@arakhmati This is an artefact of not having a consistent share file system across sites unfortunately. The instructions to download the weights are here: https://github.com/tenstorrent/tt-metal/tree/main/models/demos/wormhole/mistral7b#details but in general we should provide a way to run tests with random weights to avoid every dev having access to a copy in cases like this in which the weights themselves do not matter.

**Comment by yieldthought** [comment]
We have merged a version of this code that does not hang to main.

**Comment by eyonland** [comment]
@yieldthought , would you take a look at https://github.com/tenstorrent/tt-metal/commit/17d0e26ed4a9f8bd72d6a79d66a7e50cac3528da

I've not been able to reproduce the hang with the latest code from main.  Does it look like the changes should have caused it to hang?

**Comment by eyonland** [comment]
Thank you @yieldthought !

The stack track is showing us at:
```
void HWCommandQueue::finish() {
    ZoneScopedN("HWCommandQueue_finish");
    tt::log_debug(tt::LogDispatch, "Finish for command queue {}", this->id);
    std::shared_ptr<Event> event = std::make_shared<Event>();
    this->enqueue_record_event(event);
    if (tt::llrt::OptionsG.get_test_mode_enabled()) {
        while (this->num_entries_in_completion_q > this->num_completed_completion_q_reads) {
            if (DPrintServerHangDetected()) {
                // DPrint Server hang. Mark state and early exit. Assert in main thread.
                this->dprint_server_hang = true;
                this->set_exit_condition();
                return;
            } else if (tt::watcher_server_killed_due_to_error()) {
                // Illegal NOC txn killed watcher. Mark state and early exit. Assert in main thread.
                this->illegal_noc_txn_hang = true;
                this->set_exit_condition();
                return;
            }
        }
    } else {
        while (this->num_entries_in_completion_q > this->num_completed_completion_q_reads);
    }
}
```

This needs to be prioritized to the Metal Runtime team.

**Comment by eyonland** [comment]
This hang reminds me of the underlying code in our lockfree queue tt_metal/impl/dispatch/lock_free_queue.hpp.

Could we swap out this implementation with https://github.com/cameron314/concurrentqueue and verify whether this could be the source of the problem and verify that we do not have an issue here?

**Comment by smehtaTT** [comment]
@pgkeller @davorchap  - @eyonland  needs help from Runtime team

**Comment by tt-aho** [comment]
Have a fix pr #9273. Was a program cache issue not updating addresses.

**Comment by tt-aho** [comment]
The fix is now merged to main. Closing this issue.
